docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-.. currentmodule:: torch.sparse
+.. currentmodule:: torch
 
 .. _sparse-docs:
 
 torch.sparse
 ============
 
-.. warning::
+Torch supports sparse tensors which provide memory-efficient storage
+and processing algorithms of tensors for which the majority of
+elements have the same value (fill value). This allows saving storage
+and processing resources when the size and sparsity levels of tensors
+are high.  Otherwise, for small-sized or low-sparsity tensors using
+strided tensors is likely the most efficient approach.
 
-    This API is in beta and may change in the near future.
+.. _sparse-fill-value-docs:
 
-Torch supports sparse tensors in COO(rdinate) format, which can
-efficiently store and process tensors for which the majority of elements
-are zeros.
+Fill value
+++++++++++
 
-A sparse tensor is represented as a pair of dense tensors: a tensor
-of values and a 2D tensor of indices.  A sparse tensor can be constructed
-by providing these two tensors, as well as the size of the sparse tensor
-(which cannot be inferred from these tensors!)  Suppose we want to define
-a sparse tensor with the entry 3 at location (0, 2), entry 4 at
-location (1, 0), and entry 5 at location (1, 2).  We would then write:
+In Torch, the fill value is not specified explicitly but is implicitly
+interpreted by sparse algorithms. The implicit interpretation may vary
+in-between algorithms of different fields and applications. The most
+commonly assumed fill value is zero and it is assumed for all tools
+that are available via :mod:`torch` namespace. When some sparse tensor
+functionality assumes a non-zero or indefinite fill value, the
+corresponding tools are provided via :mod:`torch.sparse`
+namespace. Please see the documentation strings of the corresponding
+functions for a particular definition of the fill value.
 
-    >>> i = torch.LongTensor([[0, 1, 1],
-                              [2, 0, 2]])
-    >>> v = torch.FloatTensor([3, 4, 5])
-    >>> torch.sparse.FloatTensor(i, v, torch.Size([2,3])).to_dense()
-     0  0  3
-     4  0  5
-    [torch.FloatTensor of size 2x3]
+.. _sparse-hybrid-docs:
 
-Note that the input to LongTensor is NOT a list of index tuples.  If you want
+Hybrid tensors
+++++++++++++++
+
+Torch supports hybrid sparse tensors with dimensions split into sparse
+dimensions and dense dimensions. Dense dimensions always follow sparse
+dimensions, that is, mixing of dense and sparse dimensions is not
+supported. One can interpret hybrid tensors as sparse tensors with
+elements being (contiguous) strided tensors. We use (M +
+K)-dimensional tensor to denote a N-dimensional hybrid sparse tensor
+where M and K are the number of sparse and dense dimensions,
+respectively, such that M + K == N holds.
+
+See :ref:`Hybrid COO tensors <sparse-hybrid-coo-docs>`.
+
+Sparse tensor formats
++++++++++++++++++++++
+
+Torch supported sparse storage formats are:
+
+- COO(rdinate) format where tensor elements are stored using the lists
```

Comment:
Is the actual name for this format "coordinate list" or "coordinate?"

The name should be rendered "Coordinate list format, or COO format, ..." or "Coordinate format, or COO format, ..." (depending on the answer to the above).

Reply:
Here's a quick survey:
- google("coordinate format" "sparse") gives about 14 000 hits
- google("COO format" "sparse") gives about 7000 hits
- google("coordinate list format" "sparse") gives about 1000 hits
- [Wikipedia](https://en.wikipedia.org/wiki/Sparse_matrix#Coordinate_list_(COO)) defines "Coordinate list (COO)" and describes it as "COO stores a list of (row, column, value) tuples." Notice that this description does not apply to PyTorch (nor other sparse libraries) that use different storage schema (rows, columns, values are often stored as separate arrays).
- [Intel MKL sparse docs](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/appendix-a-linear-solvers-basics/sparse-matrix-storage-formats/sparse-blas-coordinate-matrix-storage-format.html) use "COO format"
- [GSL](https://www.gnu.org/software/gsl/doc/html/spmatrix.html#coordinate-storage-coo) uses "coordinate storage format"
- Magma docs use "COO format"

So, I'd suggest using "Coordinate format, or COO format, ..." in the introduction of the format and "COO format" elsewhere.

Reply:
Sounds good.